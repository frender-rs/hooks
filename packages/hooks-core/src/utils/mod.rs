use std::{ops::DerefMut, pin::Pin};

/// Re-implement [`Pin::as_deref_mut`] which is unstable for now.
///
/// [`Pin::as_deref_mut`]: https://doc.rust-lang.org/std/pin/struct.Pin.html#method.as_deref_mut
#[must_use = "`self` will be dropped if the result is not used"]
#[inline(always)]
pub fn pin_as_deref_mut<P: DerefMut>(this: Pin<&mut Pin<P>>) -> Pin<&mut P::Target> {
    // SAFETY: See Pin::as_deref_mut
    unsafe { this.get_unchecked_mut() }.as_mut()
}
