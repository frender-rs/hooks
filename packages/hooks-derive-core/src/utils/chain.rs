use darling::ToTokens;

pub struct Chain<A, B>(pub A, pub B);

impl<A, B> Chain<A, B> {
    pub fn chain<C>(self, c: C) -> Chain<Chain<A, B>, C> {
        Chain(self, c)
    }
}

impl<A: ToTokens, B: ToTokens> ToTokens for Chain<A, B> {
    #[inline]
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        A::to_tokens(&self.0, tokens);
        B::to_tokens(&self.1, tokens);
    }

    #[inline]
    fn to_token_stream(&self) -> proc_macro2::TokenStream {
        let mut tokens = A::to_token_stream(&self.0);
        tokens.extend(B::to_token_stream(&self.1));

        tokens
    }

    #[inline]
    fn into_token_stream(self) -> proc_macro2::TokenStream
    where
        Self: Sized,
    {
        let mut tokens = A::into_token_stream(self.0);
        tokens.extend(B::into_token_stream(self.1));

        tokens
    }
}
