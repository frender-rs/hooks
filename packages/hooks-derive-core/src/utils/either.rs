use quote::ToTokens;

pub enum Either<A, B> {
    A(A),
    B(B),
}

impl<A: ToTokens, B: ToTokens> ToTokens for Either<A, B> {
    #[inline]
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        match self {
            Either::A(a) => A::to_tokens(a, tokens),
            Either::B(b) => B::to_tokens(b, tokens),
        }
    }

    #[inline]
    fn to_token_stream(&self) -> proc_macro2::TokenStream {
        match self {
            Either::A(a) => A::to_token_stream(a),
            Either::B(b) => B::to_token_stream(b),
        }
    }

    #[inline]
    fn into_token_stream(self) -> proc_macro2::TokenStream
    where
        Self: Sized,
    {
        match self {
            Either::A(a) => A::into_token_stream(a),
            Either::B(b) => B::into_token_stream(b),
        }
    }
}
