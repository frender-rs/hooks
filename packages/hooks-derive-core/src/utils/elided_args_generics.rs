use syn::punctuated::Punctuated;

#[inline]
pub fn process_top_level_ref(ty: &mut syn::Type, mut f: impl FnMut(&mut syn::TypeReference)) {
    match ty {
        syn::Type::Group(g) => process_top_level_ref(&mut g.elem, f),
        syn::Type::Reference(tp) => f(tp),
        _ => {}
    }
}

/// auto fill elided lifetimes.
/// Currently only top-level references are auto filled.
///
/// For example: `&T` -> `&'hooks_auto_lifetime_0 T`
pub fn auto_fill_lifetimes(
    args_lifetimes: &mut Punctuated<syn::GenericParam, syn::Token![,]>,
    args_types: &mut Punctuated<syn::Type, syn::Token![,]>,
) {
    let mut index = 0u8;

    for ty in args_types.iter_mut() {
        process_top_level_ref(ty, |tr| {
            if tr.lifetime.is_none() {
                let lt = format!("'hooks_auto_lifetime_{index}");
                index += 1;

                let lt = syn::Lifetime::new(&lt, tr.and_token.span);
                tr.lifetime = Some(lt.clone());
                args_lifetimes.push(syn::GenericParam::Lifetime(syn::LifetimeDef::new(lt)));
            }
        })
    }
}
