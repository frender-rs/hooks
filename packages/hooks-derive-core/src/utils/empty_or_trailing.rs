use quote::ToTokens;
use syn::punctuated::Punctuated;

pub trait IsEmptyOrTrailing {
    type Punct;

    fn is_empty_or_trailing(&self) -> bool;
}

impl<T: IsEmptyOrTrailing> IsEmptyOrTrailing for &T {
    type Punct = T::Punct;

    #[inline]
    fn is_empty_or_trailing(&self) -> bool {
        T::is_empty_or_trailing(self)
    }
}

impl<T, P> IsEmptyOrTrailing for Punctuated<T, P> {
    type Punct = P;

    #[inline]
    fn is_empty_or_trailing(&self) -> bool {
        self.empty_or_trailing()
    }
}

/// [`ToTokens`] makes sure there is a trailing punct unless empty.
pub struct AutoEmptyOrTrailing<T: IsEmptyOrTrailing>(pub T);

impl<T: IsEmptyOrTrailing> IsEmptyOrTrailing for AutoEmptyOrTrailing<T> {
    type Punct = T::Punct;

    #[inline]
    fn is_empty_or_trailing(&self) -> bool {
        true
    }
}

impl<T: IsEmptyOrTrailing> ToTokens for AutoEmptyOrTrailing<T>
where
    T: ToTokens,
    T::Punct: Default + ToTokens,
{
    #[inline]
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        let e = self.0.is_empty_or_trailing();
        T::to_tokens(&self.0, tokens);
        if !e {
            let punct = T::Punct::default();
            tokens.extend(punct.into_token_stream());
        }
    }

    #[inline]
    fn to_token_stream(&self) -> proc_macro2::TokenStream {
        let e = self.0.is_empty_or_trailing();
        let mut tokens = T::to_token_stream(&self.0);
        if !e {
            let punct = T::Punct::default();
            tokens.extend(punct.into_token_stream());
        }
        tokens
    }

    #[inline]
    fn into_token_stream(self) -> proc_macro2::TokenStream
    where
        Self: Sized,
    {
        let e = self.0.is_empty_or_trailing();
        let mut tokens = T::into_token_stream(self.0);
        if !e {
            let punct = T::Punct::default();
            tokens.extend(punct.into_token_stream());
        }
        tokens
    }
}
