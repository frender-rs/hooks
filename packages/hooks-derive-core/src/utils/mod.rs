pub mod chain;
pub mod either;
pub mod elided_args_generics;
pub mod empty_or_trailing;
pub mod group;
pub mod map;
pub mod phantom;
pub mod repeat;
pub mod type_generics;
