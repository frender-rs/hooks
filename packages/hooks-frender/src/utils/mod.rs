mod option;
mod pin;
mod reentrant;

pub use option::*;
pub use pin::*;
pub use reentrant::*;

macro_rules! def_dom_element {
    (
        $vis:vis $name:ident (
            props: $name_props:ident,
            builder: $name_builder:ident,
            state: $name_state:ident,
            element: $ty_element:ty,
            props_types: $props_types:ident,
        )
        $([ $($impl_generics:tt)* ])?
        {
            $(
                $field:ident
                $(
                    $({ $($field_impl_generics_lt:tt)* })?
                    $([$field_tg:ident $($field_impl_generics:tt)* ])?
                    ($field_pat:ident : $field_prop_ty:tt) -> $field_state_ty:ty
                    { $($field_impl:tt)* }
                )?
                $(: $field_ty:ty)?
            )*
        }
    ) => {
        $vis trait $props_types $(< $($impl_generics)* >)? {
            $(
                $(type $field_tg;)?
            )*
        }

        #[non_exhaustive]
        $vis struct $name_state <$( $($impl_generics)* ,)? PropsTypes: ?Sized + $props_types> {
            pub node: ::core::option::Option<$ty_element>,
            $(
                pub $field: $($field_state_ty)? $($field_ty)? ,
            )*
        }

        $vis struct $name_props $(< $($impl_generics)* >)? {
            $(
                pub $field: $($field_prop_ty)? $($field_ty)? ,
            )*
        }

        #[non_exhaustive]
        $vis struct $name_builder $(< $($field)* >)? {
            $(
                pub $field: $($field)* ,
            )*
        }

        $vis struct $name
        $(< $($impl_generics)* >)?
        (pub $name_props < $($($field_prop_ty)? $($field_ty)?)* >);


    };
}

#[cfg(aaa)]
def_dom_element! {
    Button(
        props: ButtonProps,
        builder: ButtonPropsBuilder,
        state: ButtonState,
        element: web_sys::HtmlButtonElement,
        props_types: ButtonPropsTypes,
    ){
        // children
    }
}
