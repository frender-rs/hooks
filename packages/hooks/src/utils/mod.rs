pub mod debug_pointer;

mod rc_status;

pub use rc_status::*;

macro_rules! __impl_unexpected_token {
    () => {};
}

#[allow(unused_imports)]
pub(crate) use __impl_unexpected_token;

macro_rules! __impl_poll_next_update {
    (@ [$(#[$meta_poll:meta])*] $poll_next_update:ident [$this_0:ident $($this_1:ident)?] [$cx_pat:pat] $impl_poll:block) => {
        $(#[$meta_poll])*
        fn $poll_next_update(
            $this_0 $($this_1)? : ::core::pin::Pin<&mut Self>,
            $cx_pat: &mut ::core::task::Context<'_>,
        ) -> ::core::task::Poll<bool> $impl_poll
    };
    ($(#[$meta_poll:meta])* $poll_next_update:ident($this_0:ident $($this_1:ident)? $(,)?             ) $impl_poll:block) => {
        $crate::utils::__impl_poll_next_update! {@ [$(#[$meta_poll])*] $poll_next_update [$this_0 $($this_1)?] [ _     ] $impl_poll}
    };
    ($(#[$meta_poll:meta])* $poll_next_update:ident($this_0:ident $($this_1:ident)?, $cx_pat:pat $(,)?) $impl_poll:block) => {
        $crate::utils::__impl_poll_next_update! {@ [$(#[$meta_poll])*] $poll_next_update [$this_0 $($this_1)?] [$cx_pat] $impl_poll}
    };
    ($(#[$meta_poll:meta])* $poll_next_update:ident($($pat_args_poll:tt)*) $impl_poll:block) => {
        $(#[$meta_poll])*
        fn $poll_next_update(
            self: ::core::pin::Pin<&mut Self>,
            _: &mut ::core::task::Context<'_>,
        ) -> ::core::task::Poll<bool> $impl_poll
        $crate::utils::__impl_unexpected_token!($($pat_args_poll)*);
        compile_error!("Expect arguments for poll_next_update");
    };
    ($($tt:tt)*) => {
        fn poll_next_update(
            self: ::core::pin::Pin<&mut Self>,
            _: &mut ::core::task::Context<'_>,
        ) -> ::core::task::Poll<bool> {
            ::core::todo!()
        }
        $crate::utils::__impl_unexpected_token!($($tt)*);
        compile_error!("Expect `poll_next_update(self, cx) {}`");
    };
}

pub(crate) use __impl_poll_next_update;

macro_rules! impl_hook {
    (
        impl $([ $($impl_generics:tt)* ])? for $ty:ty $(where [$($where:tt)*])? {
            $(#[$meta_poll:meta])*
            $poll_next_update:ident($($pat_args_poll:tt)*) $impl_poll:block

            $(#[$meta_use:meta])*
            $use_hook:ident $([$($args_generics:tt)*])? ( $self_0:ident $($self_1:ident)? $(, $arg_pat:ident : $arg_ty:ty)* $(,)? ) -> $value:ty $impl_use_hook:block
        }
    ) => {
        impl<$($($impl_generics)*)?> ::hooks_core::HookPollNextUpdate for $ty $(where $($where)*)? {
            $crate::utils::__impl_poll_next_update! {
                $(#[$meta_poll])*
                $poll_next_update($($pat_args_poll)*) $impl_poll
            }
        }

        impl<
            $($($impl_generics)*)?
            $(, $($args_generics)*)?
        > ::hooks_core::Hook<($($arg_ty ,)*)> for $ty $(where $($where)*)? {
            type Value<'hook> = $value where Self: 'hook;

            $(#[$meta_use])*
            fn $use_hook<'hook>(
                $self_0 $($self_1)?: ::core::pin::Pin<&'hook mut Self>,
                ($($arg_pat ,)*): ($($arg_ty ,)*),
            ) -> Self::Value<'hook>
            where
                Self: 'hook
            $impl_use_hook
        }
    };
}
pub(crate) use impl_hook;
