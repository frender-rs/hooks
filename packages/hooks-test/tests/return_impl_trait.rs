use std::{any::Any, fmt::Display};

use futures_lite::StreamExt;
use hooks::{core::AsyncIterableHook, hook, HookExt, HookPollNextUpdateExt};

mod utils;

use utils::{assert_return_ty, hook_macro};

#[test]
fn no_hooks() {
    hook_macro! {
        #[hook]
        fn use_return() -> impl Display {
            1
        }
    }

    assert_return_ty! {
        use_return() =>
            impl ::hooks::core::Hook<()>
                + for<'hook> ::hooks::core::HookLifetime<'hook, (), Value = impl Display>
                + ::hooks::core::HookBounds<Bounds = ()>
    };

    assert_eq!(
        use_return::hook_args(),
        hooks_derive_core::HookArgs::default()
    );

    {
        let mut hook = use_return();
        assert_eq!(std::mem::size_of_val(&hook), 0);

        futures_lite::future::block_on(async {
            assert!(!hook.next_update().await);
            assert_eq!(hook.use_hook(()).to_string(), "1");
        });

        futures_lite::future::block_on(async {
            let v = hook.next_value(()).await;
            assert!(v.is_none());
        });
    }

    hook_macro! {
        #[hook]
        fn use_type_param<T: Default + Display>() -> impl Display {
            T::default()
        }
    }

    assert_return_ty! {
        use_type_param::<String>() =>
            impl ::hooks::core::Hook<()>
                + for<'hook> ::hooks::core::HookLifetime<'hook, (), Value = impl Display>
                + ::hooks::core::HookBounds<Bounds = (::core::marker::PhantomData<String>,)>
    };

    assert_eq!(
        use_type_param::hook_args(),
        hooks_derive_core::HookArgs::default()
    );

    {
        let mut hook = use_type_param::<String>();
        assert_eq!(std::mem::size_of_val(&hook), 0);

        futures_lite::future::block_on(async {
            assert!(!hook.next_update().await);
            assert_eq!(hook.use_hook(()).to_string(), "");
        });

        futures_lite::future::block_on(async {
            let v = hook.next_value(()).await;
            assert!(v.is_none());
        });
    }

    hook_macro! {
        #[hook]
        fn use_lt<'a>() -> &'a (impl Display + 'a) {
            static VALUE: i32 = 1;
            &VALUE
        }
    }

    assert_eq!(use_lt::hook_args(), hooks_derive_core::HookArgs::default());

    fn assert_use_lt<'a>() -> impl ::hooks::core::Hook<()>
           + for<'hook> ::hooks::core::HookLifetime<
        'hook,
        (),
        &'hook (&'a (),),
        Value = &'a (impl Display + 'a),
    > + ::hooks::core::HookBounds<Bounds = (&'a (),)> {
        use_lt::<'a>()
    }

    assert_use_lt();

    {
        let mut hook = use_lt();
        assert_eq!(std::mem::size_of_val(&hook), 0);

        futures_lite::future::block_on(async {
            assert!(!hook.next_update().await);
            assert_eq!(hook.use_hook(()).to_string(), "1");
        });

        futures_lite::future::block_on(async {
            let v = hook.next_value(()).await;
            assert!(v.is_none());
        });
    }
}

#[cfg(hook_macro_fail)]
#[test]
fn no_hooks_borrow_hook() {
    hook_macro! {
        #[hook]
        fn use_hook_lt() -> &'hook (impl Display + 'hook) {
            static VALUE: i32 = 2;
            &VALUE
        }
    }

    assert_return_ty! {
        use_hook_lt() => impl ::hooks::core::Hook<()>
            + for<'hook> ::hooks::core::HookLifetime<'hook, (), Value = &'hook impl Display>
            + ::hooks::core::HookBounds<Bounds = ()>
    };

    assert_eq!(
        use_hook_lt::hook_args(),
        hooks_derive_core::HookArgs::default()
    );

    {
        let mut hook = use_hook_lt();
        assert_eq!(std::mem::size_of_val(&hook), 0);

        futures_lite::future::block_on(async {
            assert!(!hook.next_update().await);
            assert_eq!(hook.use_hook(()), &2);
        });

        futures_lite::future::block_on(async {
            let v = hook.next_value(()).await;
            assert!(v.is_none());
        });
    }
}

#[cfg(hook_macro_fail)]
#[test]
fn one_hook() {
    hook_macro! {
        #[hook]
        fn use_one_hook() -> &'hook mut impl Display {
            let v = ::hooks::use_lazy_pinned(0);
            v.get_mut()
        }
    }

    assert_return_ty! {
        use_one_hook() => impl ::hooks::core::Hook<()>
            + for<'hook> ::hooks::core::HookLifetime<'hook, (), Value = &'hook mut i32>
            + ::hooks::core::HookBounds<Bounds = ()>
    };

    assert_eq!(
        use_one_hook::hook_args(),
        hooks_derive_core::HookArgs::default()
    );

    {
        let mut hook = use_one_hook();
        assert_eq!(
            std::mem::size_of_val(&hook),
            std::mem::size_of::<::hooks::LazyPinned<i32>>()
        );

        futures_lite::future::block_on(async {
            assert!(!hook.next_update().await);
            assert_eq!(*hook.use_hook(()), 0);
            assert!(!hook.next_update().await);
            *hook.use_hook(()) = -3;
            assert!(!hook.next_update().await);
            assert_eq!(*hook.use_hook(()), -3);
        });

        futures_lite::future::block_on(async {
            let v = hook.next_value(()).await;
            assert!(v.is_none());
        });
    }
}

#[test]
fn one_state() {
    hook_macro! {
        #[hook]
        fn use_str_state() -> impl Any {
            let (state, updater) = ::hooks::use_state_with(String::new);

            updater.replace_maybe_with_fn_pointer(|old| {
                if old.len() < 2 {
                    Some(format!("{old} "))
                } else {
                    None
                }
            });

            state.clone()
        }
    }

    assert_return_ty! {
        use_str_state() => impl ::hooks::core::Hook<()>
            + for<'hook> ::hooks::core::HookLifetime<'hook, (), Value = impl Any>
            + ::hooks::core::HookBounds<Bounds = ()>
    };

    assert_eq!(
        use_str_state::hook_args(),
        hooks_derive_core::HookArgs::default()
    );

    fn assert_string<T: Any>(t: &T) -> &String {
        <dyn Any>::downcast_ref::<String>(t).unwrap()
    }

    {
        let mut hook = use_str_state();
        assert_eq!(
            std::mem::size_of_val(&hook),
            std::mem::size_of::<::hooks::State<String>>()
        );

        futures_lite::future::block_on(async {
            assert!(hook.next_update().await);
            assert_eq!(assert_string(&hook.use_hook(())), "");
            assert!(hook.next_update().await);
            assert_eq!(assert_string(&hook.use_hook(())), " ");
            assert!(hook.next_update().await);
            assert_eq!(assert_string(&hook.use_hook(())), "  ");
            assert!(!hook.next_update().await);
        });
    }

    futures_lite::future::block_on(async {
        let values = use_str_state().into_iter().collect::<Vec<_>>().await;

        let values = values
            .iter()
            .map(assert_string)
            .map(String::as_str)
            .collect::<Vec<_>>();

        assert_eq!(values, ["", " ", "  "])
    });
}

#[test]
fn two_hooks() {
    hook_macro! {
        #[hook]
        fn use_state_effect() -> impl Display {
            let (state, updater) = ::hooks::use_state_with::<i32>(Default::default);
            let updater = updater.clone();

            ::hooks::use_effect(move |v: &_| {
                if *v < 2 {
                    updater.set(*v + 1)
                }
            }, *state);

            *state
        }
    }

    assert_return_ty! {
        use_state_effect() => impl ::hooks::core::Hook<()>
            + for<'hook> ::hooks::core::HookLifetime<'hook, (), Value = impl Display>
            + ::hooks::core::HookBounds<Bounds = ()>
    };

    assert_eq!(
        use_state_effect::hook_args(),
        hooks_derive_core::HookArgs::default()
    );

    {
        let mut hook = use_state_effect();

        let dummy_effect = {
            let updater = hooks::StateUpdater::<i32>::new();
            let effect = move |_: &i32| drop(updater);
            let mut hook = hooks::use_effect();
            hook.use_hook((effect, 0));
            hook
        };

        assert_eq!(
            std::mem::size_of_val(&hook),
            std::mem::size_of::<::hooks::State<i32>>() + std::mem::size_of_val(&dummy_effect),
        );

        futures_lite::future::block_on(async {
            assert!(hook.next_update().await);
            assert_eq!(hook.use_hook(()).to_string(), "0");
            assert!(hook.next_update().await);
            assert_eq!(hook.use_hook(()).to_string(), "1");
            assert!(hook.next_update().await);
            assert_eq!(hook.use_hook(()).to_string(), "2");
            assert!(!hook.next_update().await);
        });
    }

    futures_lite::future::block_on(async {
        let values = use_state_effect()
            .into_iter()
            .map(|v| v.to_string())
            .collect::<Vec<_>>()
            .await;

        assert_eq!(values, ["0", "1", "2"]);
    });
}
