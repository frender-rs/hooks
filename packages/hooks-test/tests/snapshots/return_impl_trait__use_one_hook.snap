---
source: packages/hooks-test/tests/return_impl_trait.rs
expression: "use_one_hook::hook_impl_code()"
---
fn use_one_hook() -> impl ::hooks::core::UpdateHookUninitialized<
    Uninitialized = impl ::hooks::core::HookPollNextUpdate + ::hooks::core::HookUnmount + ::core::default::Default,
    Hook = impl ::hooks::core::Hook + for<'hook> ::hooks::core::HookValue<
        'hook,
        Value = &'hook mut (impl Display + Default),
    >,
> {
    enum __HooksImplNever {}
    struct __HooksValueOfThisHook<HooksImplTrait0: Display + Default> {
        __: (__HooksImplNever, ::core::marker::PhantomData<HooksImplTrait0>),
    }
    impl<'hook, HooksImplTrait0: Display + Default> ::hooks::core::HookValue<'hook>
    for __HooksValueOfThisHook<HooksImplTrait0> {
        type Value = &'hook mut HooksImplTrait0;
    }
    ::hooks::core::fn_hook::use_fn_hook::<
        __HooksValueOfThisHook<_>,
        _,
        _,
    >(move |__hooks_hook_0: ::core::pin::Pin<&mut _>| {
        let v = ::hooks::core::UpdateHookUninitialized::h(
            ::hooks::use_lazy_pinned(-3),
            __hooks_hook_0,
        );
        v.get_mut()
    })
}
